Refactoring api endpoints

Lruquaf opened the pull request:
- register routes were modified according to MSC model. 
- **user.controller.js**, **user.service.js** files were created. 
- **User.js** model filename was changed with **user.model.js**. 
- **register.js** route filename was changed with **register.route.js**. 
-  _isAuthenticated_ in **auth.js** was fixed. **verifySignature.js** and **checkSessionToken.js** middlewares were created. 
- **sessionHelper.js** file in helpers folder was created to handle session interactions. 
- In _register/dev_ route, "email" section was removed due to unnecessity. In this way, _register/dev_ route can use _registerUser_ function in **user.controller.js**. 
- It is confused that _signerClient.verifyMessage_ function is async, or not. Both are working but if this function is marke as async, then _verifySignature_ middleware must be an async function just like in the code. Is this a problem that a middleware is an async?

BlocksOnChain commented:
@Lruquaf If verifySignature is not a promise, you don't have to use async await.

BlocksOnChain commented on controllers/user.controller.js:
```diff
@@ -0,0 +1,49 @@
+
+async function registerUser(req, res) {
+	const { walletAddress } = req.body;
+	try {
```
You can make the if else blocks into just one function as registerOrCreateUser() which should use two separate functions, one of them should be what if does there, and the other one is the else part. You gonna call them from registerOrCreateUser function. they all gonna be inside services and controller will call it as one function inside try catch

BlocksOnChain commented:
I think also you need to remove register as a route. Because it is actually users route. Lets make the name of the route as users/register and use register routes inside user.route.

Lruquaf commented:
## Changes With Last Commit
### Routes
- **register.route.js** was merged into **user.route.js** and removed from code.
- Endpoints in **user.route.js** were refactored according to MSC pattern. (Q1: Should any control or middleware be in GET /users endpoint? Q2: Should /users/put/email endpoint be POST or PUT?)
### Controller
- _get-message-to-sign_ function edited, its token and message creation functionality were moved to _createTokenAndMessage_ function in **sessionHelper.js**.
- _registerUser_ function was refactored with a new function named _registerOrCreateUser_ in its content.
- _getAllUsers_ and _putEmail_ functions were created.
### Service
- User existence control was added in _findUserByWalletAddress_ function. (Q: Is this error throwing approach ok or not? Is "return res.status(401)" more proper? If yes, how and where can I implement it?
- _findUserById_ and _findAllUsers_ functions were created.
- _registerOrCreateUser_ function was created and divided into two functions named _findAndRegisterUser_ and _createAndRegisterUser_ to control user existence and set session.
- _verifySignature_ function was moved in **helperFunctions.js** in helpers folder.
### Helper Functions
- **helperFunctions.js** in helpers folder was created.
- _verifySignature_ and _validateEmail_ functions were removed **helperFunctions.js**.
### Session Helper
- _createTokenAndMessage_ function was added. It creates session token and session message.
### Middlewares
- **verifySignature.js** and **isEmailValid** middlewares were created and implemented in routes. (Q1: Naming middleware files and function seem a bit weird. Is there any best practice or suggestion? Q2: Is file structure of middlewares ok?)

BlocksOnChain commented on services/user.service.js:
```diff
@@ -0,0 +1,72 @@
+	if (user.length === 0) {
+		return await createAndRegisterUser(req, walletAddress);
+	} else {
+		return await findAndRegisterUser(req, walletAddress);
```
This should have the name findAndLogin and the other should be createAndRegister, the reason you should not put User at the end of these is because you are already inside user.service. So you should always use these functions as UserService.createAndRegister() etc...

BlocksOnChain commented:
q1: Currently we don't have admins but I think it will be better if you add admin type user. And add middleware that checks for only admin
q2: It can be post for now 
Services q1: You should return the function return of service in the controller and inside controller you would have try catch, when you catch the error you should return 500 with error message.
Middleware q: using validate or verify prefix is ok for middleware, also **is** prefix is widely used. You can ask an LLM for better naming if you want though

BlocksOnChain commented:
Also I see your point. I think putting all middleware functions into one place can be done.

Lruquaf commented:
## Changes With Last Commit
### Middleware
- All middlewares were renamed and merged into **middleware.js** file in middleware folder.
### Routes
- Middlewares were implemented to routes properly.
### Controller and Service
- _registerOrCreateUser_ and associated functions were renamed.
### Helper Functions
- A bug in _verifySignature_ function was fixed. Now, signing message and verifying signature flow are working well.

Lruquaf commented:
## Changes With Commit [0fc746b](https://github.com/Esayf/Examina-Backend-API/pull/21/commits/0fc746bba0dc8b3fd7477a44a7a3b9989607f5de)
### Controllers
- **exam.controller.js** was created.
- _createExam_, _getAllExams_ and _getExamById_ functions were created. (Q: Session user control in line 47 seems unnecessary? What is the purpose of that?)
### Helper Functions
- Some functions related to extracting CIDs and pinning to IPFS were added.
### Models
- Exam model was renamed.
### Routes
- **exam.route.js** was created for exam routes.
### Services
- Some functions related to exam routes like _create_ and _saveQuestions_ were created. (Q: In _getParticipation_ function, I think there is an unnecessary control of "userId" existence on line 93. I have commented it out.)
- Protokit interation was removed from the exam functions.
- Error catching mechanism was added to the functions in **user.service.js**.
- Functions naming in **user.service.js** was fixed.

BlocksOnChain commented on controllers/exam.controller.js:
```diff
@@ -0,0 +1,63 @@
+		const result = await examService.getByIdWithParticipation(
+			req.params.id,
+			// Unnecessary Control?
+			req.session.user ? req.session.user.userId : null
```
This is a control to check if user has session. It would throw error on req.session.user.userId so you would have to use req.sessions.user?.userId .  That would shorten that line but necessary

Lruquaf replied:
Yeah, got it. But We have an authorization check from middleware which is used for these routes already. So, I think we don't need this control.

ClockChainXD replied:
In the past getExamById could return exam details without user session, that might be handy for us for future. It should return only exam if no session, return isFinished if there is session. But you can remove the unnecessary check for getParticipation service function

Lruquaf replied:
I understood your approach. I will implement it but our older version auth control was there for this route, and it was working interestingly. anyway, I m gonna do it.

BlocksOnChain replied:
Oh then there was a weird mistake there that we forgot, ok cool thx good job  😁

Lruquaf commented:
## Changes With Commit [796ac11](https://github.com/Esayf/Examina-Backend-API/pull/21/commits/796ac1184a3d700493120c1823935d27caa63b5f)
### Routes
- _/startExam_ and _/:id/questions_ routes were added in **exam.route.js**. (Q1: Should _/:id/questions_ route be in another route file like "questions.route.js"? Because we get questions of an exam by that. Q2: Is _ensureAuthenticated_ ok for this endpoint?)
- _ensureAuthenticated_ middleware was put on proper routes separately.)
### Controllers
- _startExam_ and _getExamQuestions_ were created.
### Services
#### exam.service
- _start_ and _getQuestionsByExam_ functions were created. (Q1: should this function move to **question.service**. It would be better maybe. Q2: These controls in these functions like _checkExamTimes_ should be inside service or controller? Which one is more proper?)
- Some minor fixes were added in other functions.
#### participatedUser.service
- **participatedUser.service** file was created and functions associated with **ParticipatedUser** model were created.
- _checkParticipation_ functions related to getting questions and starting exam logics were merged and implemented in a single function named _checkParticipation_ in this file.
#### question.service
- The file was created with function named _getAllByExam_.
### Helper Functions
- _checkExamTimes_ functions was added.

Lruquaf commented:
## Changes With Commit [110bfaa](https://github.com/Esayf/Examina-Backend-API/pull/21/commits/110bfaa983f025c425ae6869efdfacc9fa45525b)

(Please answer the questions in code comments, If you want we can have a meet about questions or changes or other recommendations.)

### Controllers
- **answer.controller.js**, **question.controller.js**, **score.controller.js** files were created and associated functions added.
- Some error and handling lines were fixed in **exam.controller.js** and **user.controller.js** files.
- _finishExam_ function was added in **exam.controller.js**.
### Cron
- **checkCompletedExams.js** file was created and protokit functions were removed.
### Helper Functions
- _generateAnswerArray_ and _isExamCompleted_ functions were added.
### Middleware
- _validateBody_ function from **examValidators.js** file was created and applied to finishExam route.
- **protokit.js** file was removed.
### Models
- Some files were renamed.
### Routes
- **answer.route.js**, **score.route.js**, **question.route.js** were created and associated routes were added.
- _finishExam_ route was created.
- Old route files were removed.
### Services
- **answer.service.js** and **score.service.js** files were created.
- Some error messages and naming issues were fixed in **user.service.js**, **participatedUser.service.js**, **question.service.js** and **answer.service.js** files.
- Finish exam logic was added in **exam.service.js**.

BlocksOnChain commented on services/exam.service.js:
```diff
@@ -101,31 +126,38 @@ async function start(examId, userId) {
-// This function should move to question.service. It would be better.
-// These controls should be inside service or controller? Which one is more proper?
+// Q: This function should move to question.service. It would be better.
+// Q: These controls should be inside service or controller? Which one is more proper?
```
Validation of the id is ok to do in the services, normally what we do as a pattern here is to use decorators, which actually a middleware that checks the id in the parameter, like you do in if(!exam) part and returns full exam or anything that is needed as DTO. But it is not that important here. Those are for big projects

BlocksOnChain commented on services/exam.service.js:
```diff
@@ -101,31 +126,38 @@ async function start(examId, userId) {
 
-// This function should move to question.service. It would be better.
-// These controls should be inside service or controller? Which one is more proper?
+// Q: This function should move to question.service. It would be better.
```
Yes it should be in questions service
